x-pack/plugins/reporting/server/export_types/csv_searchsource/generate_csv/generate_csv.ts

```diff
@@ -242,26 +242,35 @@ export class CsvGenerator {
   /*
    * Format a Datatable into rows of CSV content
    */
-  private generateRows(
+  private async generateRows(
     columns: string[],
     table: Datatable,
     builder: MaxSizeStringBuilder,
     formatters: Record<string, FieldFormat>,
     settings: CsvExportSettings
   ) {
     this.logger.debug(`Building ${table.rows.length} CSV data rows...`);
+
+    const asyncGenerateRow = async (dataTableRow: Record<string, any>): Promise<string> => {
+      return new Promise((resolve) => {
+        setImmediate(() => {
+          resolve(
+            columns
+              .map((f) => ({ column: f, data: dataTableRow[f] }))
+              .map(this.formatCellValues(formatters))
+              .map(this.escapeValues(settings))
+              .join(settings.separator) + '\n'
+          );
+        });
```

Comment:
We can get more performance gains here by doing this instead:

```ts
          const row: string[] = [];
          const format = this.formatCellValues(formatters);
          const escape = this.escapeValues(settings);

          for (const column of columns) {
            row.push(escape(format({ column, data: dataTableRow[column] })));
          }
          resolve(row.join(settings.separator) + '\n');
```

In my testing I saw this took up to half the time to execute per call (just doing end-start on Date.now() values)

before: `csv.generateData.generateRows took 0.576 second`
after: `csv.generateData.generateRows took 0.247 seconds`

Reply:
https://github.com/elastic/kibana/pull/120309/commits/ebacdbbb0640cc0b6577475dbc9be3b814b8d99d